iroh/src/magicsock.rs

```diff
@@ -2421,6 +2525,87 @@ impl Actor {
         self.update_direct_addresses(report);
     }
 
+    /// Does a DNS look up for the `RelayUrl` and returns the set of resolved
+    /// [`SocketAddr`]s
+    async fn resolve_relay_quic_endpoint(
+        dns_resolver: DnsResolver,
+        relay_node: Arc<RelayNode>,
+        duration: Duration,
+    ) -> BTreeSet<SocketAddr> {
+        let mut addrs = BTreeSet::new();
+        let port = if let Some(ref quic_config) = relay_node.quic {
+            quic_config.port
+        } else {
+            trace!(
+                ?relay_node,
+                "No quic config for the relay node: no need to resolve quic endpoint ip"
+            );
+            return addrs;
+        };
+
+        match relay_node.url.host() {
+            Some(url::Host::Domain(hostname)) => {
+                debug!(%hostname, "Performing DNS A lookup for relay addr");
+                let mut set = JoinSet::new();
+                let resolver = dns_resolver.clone();
+                let ipv4_resolver = resolver.clone();
+                let ipv4_hostname = hostname.to_owned();
+                set.spawn(async move {
+                    let res = ipv4_resolver.lookup_ipv4(ipv4_hostname, duration).await;
+                    res.map(|addrs| addrs.collect::<Vec<_>>())
+                });
+                let ipv6_hostname = hostname.to_owned();
+                set.spawn(async move {
+                    let res = resolver.lookup_ipv6(ipv6_hostname, duration).await;
+                    res.map(|addrs| addrs.collect::<Vec<_>>())
+                });
+                let responses = set.join_all().await;
+                for res in responses {
+                    match res {
+                        Err(_) => {}
+                        Ok(resolved_addrs) => {
+                            for addr in resolved_addrs {
+                                addrs.insert(SocketAddr::new(addr, port));
+                            }
+                        }
+                    }
+                }
+                if addrs.is_empty() {
+                    debug!(%hostname, "Unable to resolve ip addresses for relay node");
+                }
+            }
+            Some(url::Host::Ipv4(addr)) => {
+                addrs.insert(SocketAddr::new(addr.into(), port));
+            }
+            Some(url::Host::Ipv6(addr)) => {
+                addrs.insert(SocketAddr::new(addr.into(), port));
+            }
+            None => {
+                error!(?relay_node.url, "No hostname for relay node, cannot resolve ip");
+            }
+        }
+        addrs
+    }
+
+    /// Resolve the relay addresses used for QUIC address discovery.
+    async fn resolve_qad_addrs(&mut self, duration: Duration) {
+        let mut set = JoinSet::new();
```

Comment:
Probably a https://docs.rs/futures-buffered/latest/futures_buffered/struct.FuturesUnorderedBounded.html is better suited than a set of tasks?  Not sure how big a difference it makes to be fair.  But it seems a little more elegant.

(and also not sure if this code will end up staying around or not)